js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts

```diff
@@ -0,0 +1,591 @@
+    model: ViewModelConstructor<O>,
+    fields: O
+): ['id', NumberField] {
+    return ['id', fields.id || new NumberField()];
+}
+
+// Overloads here are so we can more accurately type the primary key and fields (specifically for default case we can
+// add the implicit 'id' field that will be created)
+export default function ViewModelFactory<O extends FieldsMapping>(
+    options: ViewModelOptionsPkFieldNameSingle<O>
+): ViewModelConstructor<O, string, SinglePrimaryKey>;
+export default function ViewModelFactory<O extends FieldsMapping>(
+    options: ViewModelOptionsPkFieldNameCompound<O>
+): ViewModelConstructor<O, string[], CompoundPrimaryKey>;
+export default function ViewModelFactory<O extends FieldsMapping>(
+    options: ViewModelOptionsGetImplicitPkFieldSingle<O>
+): ViewModelConstructor<O, string, SinglePrimaryKey>;
+export default function ViewModelFactory<O extends FieldsMapping>(
+    options: ViewModelOptionsGetImplicitPkFieldCompound<O>
+): ViewModelConstructor<O, string[], CompoundPrimaryKey>;
+export default function ViewModelFactory<O extends FieldsMapping>(
+    options: ViewModelOptions<O>
+): ViewModelConstructor<{ id: NumberField } & O, 'id', string | number>;
+export default function ViewModelFactory<O extends FieldsMapping>(
+    options: ViewModelOptions<O>
+): ViewModelConstructor<any, any> {
+    const { fields } = options;
+    if (options.pkFieldName && options.getImplicitPkField) {
+        throw new Error("Only one of 'pkFieldName' and 'getImplicitPkField' should be provided");
+    }
+    // If pkFieldName isn't specified it will be created automatically as 'id'
+    // Otherwise the field name will be included in `fields` and we don't need to modify the type.
+    // getImplicitPkField will create a new field if specified but we can't type it so we ignore it
+    // (nested ternary is unavoidable to use the typescripts `extends` behaviour)
+    type FinalFields = typeof options.pkFieldName extends string | string[]
+        ? O
+        : typeof options.getImplicitPkField extends Function
+        ? O
+        : { id: NumberField } & O;
+    type PkFieldType = typeof options.pkFieldName extends string
+        ? string
+        : typeof options.pkFieldName extends string[]
+        ? string[]
+        : typeof options.getImplicitPkField extends GetImplicitPkFieldCompound<O>
+        ? string[]
+        : typeof options.getImplicitPkField extends GetImplicitPkFieldSingle<O>
+        ? string
+        : 'id';
+
+    type PkValueType = typeof options.pkFieldName extends string
+        ? SinglePrimaryKey
+        : typeof options.pkFieldName extends string[]
+        ? CompoundPrimaryKey
+        : typeof options.getImplicitPkField extends GetImplicitPkFieldCompound<O>
+        ? CompoundPrimaryKey
+        : typeof options.getImplicitPkField extends GetImplicitPkFieldSingle<O>
+        ? SinglePrimaryKey
+        : SinglePrimaryKey;
+
+    // This is the constructor function for the created class
```

Comment:
Why is this not just a regular `constructor()` on a regular class?

ie:

```
function ViewModelFactory() {

   class GenericViewModel {
      constructor() {
          ...
      }
   }

  GenericViewModel.name = ...;

  return GenericViewModel;
}
```

Reply:
couldn't get type definitions to work. i think particularly the indexing signature